### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by oscarduignan**
Originally it was a bash script, then it was a small python script, then rather than rewrite in Scala I refined the python a bit which seemed the most interoperable with dependencies and related tools like datasette which can be used for visualising and platops' "checkout all repositories" script which can be used to get a copy of all codebases to search against.

- searches are done using rg and jq then streams of results are processed by python
- python combines them and streams them to an output (stdout, file, or sqlite db)
- output can then be used by things like sqlite-utils to query the data with SQL or JQ to filter and transform and remix it further, or datasette to spin up a web interface to query, facet, and visualize the results (screenshot in readme)
- takes a bit of time to complete when searching a large number of frontends because it runs a lot of subprocesses, on my 8 core laptop it is around 180 seconds when searching 254 repositories; which have 11886 templates; totalling 682,384 lines and finding 8121 usages
- because it's using regex the results aren't perfect, but some classification is attempted to show when a usage is included via a certain mechanism like dependency injection or has used our `withFormField` helpers inline
- there are some fixtures with expected output that can check the core behaviour with pytest but there's no coverage for the cli tools or outputs at the moment
- no source code or actual output is included in this repository, you need to provide that yourself and not publish the outputs when run against potentially private repositories

### CODE REVIEW

File: datasette/plugins/render_code_sample.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+        return jinja2.Markup(
+            f"""<a style="display:flex;align-items:center;" href="{column_json['github_url']}">{column_json['path']}#L{column_json['line_number']} <img style="height:20px" src="data:image/png;base64, {github_logo_base64}" alt=""/></a><pre class="code-sample" style="counter-reset: line {column_json['line_number'] - 1};">{escaped_code}</pre>"""
```
Comment (oscarduignan) [inline datasette/plugins/render_code_sample.py]: this was bodged in to improve the look of source code column when viewed from datasette, so it's got inline styles and is all on one line. It does the job well enough for the current scope, clicking the link takes you to the repository with the usage highlighted so you can check if it's got anything wrong / explore further

CleanShot 2021-07-19 at 09 48 28@2x

File: datasette/static/extra_styles.css [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+pre.code-sample>code:before {
+    counter-increment: line;
+    content: counter(line);
+    border-right: 1px solid #aaaaaa;
+    padding-right: .5em;
+    margin-right: .5em;
+    padding-left: .25em;
+    color: #999;
+    width: fit-fontent(3ch);
```
Comment (oscarduignan) [inline datasette/static/extra_styles.css]: I will fix this though 🤦 
```suggestion
    width: 3ch;
```

File: shell.nix [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+mkShell {
+    buildInputs = [
+        python39
+        python39Packages.poetry
+        ripgrep
+        jq
+        datamash
+        findutils
+    ];
+}
```
Comment (oscarduignan) [inline shell.nix:18]: shows you the system dependencies you need and if you have nix also lets you setup a local shell with them all available by running `nix-shell`

File: datasette/plugins/render_code_sample.py (hunk 1 above)
Comment (oscarduignan) [reply datasette/plugins/render_code_sample.py]: refactored this inline stuff into a separate template!

**Review by matthewmascord — APPROVED** [review]

**Pull request merged**